yrs/src/types/xml.rs

```diff
@@ -109,421 +132,533 @@ impl XmlElement {
     }
 }
 
-impl<'a, B, T> Iterator for Attributes<'a, B, T>
+impl<I, T> XmlPrelim for XmlFragmentPrelim<I, T>
 where
-    B: Borrow<T>,
-    T: ReadTxn,
+    I: IntoIterator<Item = T>,
+    T: XmlPrelim,
 {
-    type Item = (&'a str, String);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let (key, block) = self.0.next()?;
-        let txn = self.0.txn.borrow();
-        let value = block
-            .content
-            .get_last()
-            .map(|v| v.to_string(txn))
-            .unwrap_or(String::default());
+    type Return = XmlFragmentRef;
+}
 
-        Some((key.as_ref(), value))
+impl<I, T> Prelim for XmlFragmentPrelim<I, T>
+where
+    I: IntoIterator<Item = T>,
+    T: XmlPrelim,
+{
+    fn into_content(self, _txn: &mut TransactionMut) -> (ItemContent, Option<Self>) {
+        let inner = Branch::new(TYPE_REFS_XML_FRAGMENT, None);
+        (ItemContent::Type(inner), Some(self))
     }
-}
 
-impl Into<XmlElement> for XmlFragment {
-    fn into(self) -> XmlElement {
-        XmlElement(self)
+    fn integrate(self, txn: &mut TransactionMut, inner_ref: BranchPtr) {
+        let xml = XmlFragmentRef::from(inner_ref);
+        for value in self.0 {
+            xml.push_back(txn, value);
+        }
     }
 }
 
-#[repr(transparent)]
+/// (Obsolete) an Yjs-compatible XML node used for nesting Map elements.
 #[derive(Debug, Clone, Eq, PartialEq)]
-pub struct XmlFragment(BranchPtr);
+pub struct XmlHookRef(BranchPtr);
 
-unsafe impl Send for XmlFragment {}
-unsafe impl Sync for XmlFragment {}
+impl Map for XmlHookRef {}
 
-impl XmlFragment {
-    pub fn new(inner: BranchPtr) -> Self {
-        XmlFragment(inner)
+impl ToJson for XmlHookRef {
+    fn to_json<T: ReadTxn>(&self, txn: &T) -> Any {
+        let map: MapRef = self.clone().into();
+        map.to_json(txn)
     }
+}
 
-    fn inner(&self) -> BranchPtr {
-        self.0
+impl AsRef<Branch> for XmlHookRef {
+    fn as_ref(&self) -> &Branch {
+        self.0.deref()
     }
+}
 
-    pub fn first_child(&self) -> Option<Xml> {
-        let inner = self.inner();
-        let first = inner.first()?;
-        match &first.content {
-            ItemContent::Type(c) => {
-                let value = Xml::from(BranchPtr::from(c));
-                Some(value)
-            }
-            _ => None,
-        }
+impl AsMut<Branch> for XmlHookRef {
+    fn as_mut(&mut self) -> &mut Branch {
+        self.0.deref_mut()
     }
+}
 
-    pub fn parent(&self) -> Option<XmlElement> {
-        parent(self.inner())
+impl From<BranchPtr> for XmlHookRef {
+    fn from(inner: BranchPtr) -> Self {
+        XmlHookRef(inner)
     }
+}
 
-    pub fn len<T: ReadTxn>(&self, txn: &T) -> u32 {
-        self.inner().len()
+impl Into<MapRef> for XmlHookRef {
+    fn into(self) -> MapRef {
+        MapRef::from(self.0)
     }
+}
 
-    pub fn iter<'a, T: ReadTxn>(&'a self, txn: &'a T) -> TreeWalker<'a, &'a T, T> {
-        TreeWalker::new(&self.0, txn)
+pub trait Xml: AsRef<Branch> {
```

Comment:
`Xml` trait API is responsible for managing xml node attributes - it's implemented by `XmlTextRef` and `XmlElementRef` (but NOT by `XmlFragmentRef`).